Geometry/VeryForwardGeometryBuilder/plugins/PPSGeometryESProducer.cc

```diff
@@ -0,0 +1,177 @@
+#include "Geometry/VeryForwardGeometryBuilder/plugins/PPSGeometryESProducer.h"
+
+#include "DataFormats/CTPPSDetId/interface/TotemRPDetId.h"
+#include "DataFormats/CTPPSDetId/interface/TotemTimingDetId.h"
+#include "DataFormats/CTPPSDetId/interface/CTPPSPixelDetId.h"
+#include "DataFormats/CTPPSDetId/interface/CTPPSDiamondDetId.h"
+
+#include "DetectorDescription/DDCMS/interface/DDCompactView.h"
+#include "Geometry/VeryForwardGeometryBuilder/interface/DetGeomDescBuilder.h"
+#include "Geometry/VeryForwardGeometryBuilder/interface/CTPPSDDDNames.h"
+
+#include <regex>
+
+/**
+ * \brief Builds ideal, real and misaligned geometries.
+ *
+ * First, it creates a tree of DetGeomDesc from DDCompView. For real and misaligned geometries,
+ * it applies alignment corrections (RPAlignmentCorrections) found in corresponding ...GeometryRecord.
+ *
+ * Second, it creates CTPPSGeometry from DetGeoDesc tree.
+ **/
+
+using RotationMatrix = ROOT::Math::Rotation3D;
+using Translation = ROOT::Math::DisplacementVector3D<ROOT::Math::Cartesian3D<double>>;
+
+PPSGeometryESProducer::PPSGeometryESProducer(const edm::ParameterSet& iConfig)
+    : verbosity_(iConfig.getUntrackedParameter<unsigned int>("verbosity")),
+      detectorToken_{setWhatProduced(this, &PPSGeometryESProducer::produceIdealGD)
+                         .consumes<cms::DDDetector>(edm::ESInputTag("" /*optional module label */,
+                                                                    iConfig.getParameter<std::string>("detectorTag")))},
+      gdRealTokens_{setWhatProduced(this, &PPSGeometryESProducer::produceRealGD)},
+      gdMisTokens_{setWhatProduced(this, &PPSGeometryESProducer::produceMisalignedGD)},
+      dgdRealToken_{
+          setWhatProduced(this, &PPSGeometryESProducer::produceRealTG).consumes<DetGeomDesc>(edm::ESInputTag())},
+      dgdMisToken_{
+          setWhatProduced(this, &PPSGeometryESProducer::produceMisalignedTG).consumes<DetGeomDesc>(edm::ESInputTag())} {
+}
+
+void PPSGeometryESProducer::fillDescriptions(edm::ConfigurationDescriptions& descriptions) {
+  edm::ParameterSetDescription desc;
+  desc.addUntracked<unsigned int>("verbosity", 1);
+  desc.add<std::string>("detectorTag", std::string());
+  descriptions.add("PPSGeometryESProducer", desc);
+}
+
+//----------------------------------------------------------------------------------------------------
+/*
+ * Apply alignments by doing a BFS on idealGD tree.
+ */
+void PPSGeometryESProducer::applyAlignments(const DetGeomDesc& idealGD,
```

Comment:
It would be nice if this method could return `DetGeomDesc  *` to make it more clear it is providing this object.